src/main/java/htsjdk/samtools/BAMFileReader.java

```diff
@@ -796,59 +804,160 @@ public SAMRecord next() {
+            private int recordLength;
+            private byte[] buffer;
+            private final long start;
+            private Long stop;
+            private final long recordIndex;
+            private SAMRecord record;
+            public BamRecordDecodingInfo(long recordIndex, int recordLength, long start) {
+                this.recordIndex = recordIndex;
+                this.recordLength = recordLength;
+                this.start = start;
+            }
+
+            public BamRecordDecodingInfo(long recordIndex, int recordLength, long start, long stop, byte[] buffer) {
+                this.recordIndex = recordIndex;
+                this.recordLength = recordLength;
+                this.start = start;
+                this.stop = stop;
+                this.buffer = buffer;
+            }
+        }
+        private class AsyncBamDecoder extends AsyncReadTaskRunner<BamRecordDecodingInfo, BamRecordDecodingInfo> {
+            public AsyncBamDecoder(int batchBufferBudget, int totalBufferBudget) {
+                super(batchBufferBudget, totalBufferBudget);
+            }
+
+            @Override
+            public Tuple<BamRecordDecodingInfo, Long> performReadAhead(long bufferBudget) {
+                final BamRecordDecodingInfo record = readNextRecord();
+                if (record == null) {
+                    return new Tuple<>(null, 0L);
+                } else {
+                    return new Tuple<>(record, (long)record.recordLength);
+                }
+            }
+
+            @Override
+            public BamRecordDecodingInfo transform(BamRecordDecodingInfo record) {
+                decode(record);
+                return record;
+            }
+        }
 
         /**
          * Read the next record from the input stream.
          */
-        SAMRecord getNextRecord() throws IOException {
-            final long startCoordinate = mCompressedInputStream.getFilePointer();
-            final SAMRecord next = bamRecordCodec.decode();
-            final long stopCoordinate = mCompressedInputStream.getFilePointer();
-
-            if(mReader != null && next != null)
-                next.setFileSource(new SAMFileSource(mReader,new BAMFileSpan(new Chunk(startCoordinate,stopCoordinate))));
+        private BamRecordDecodingInfo readNextRecord() {
+            try {
+                if (!advanceToNextRecordStart()) {
+                    // effectively end of stream
+                    return null;
+                }
+                final long startCoordinate = mCompressedInputStream.getFilePointer();
+                Integer readLength = bamRecordCodec.decodeRecordLength();
+                if (readLength == null) {
+                    // EOF returns null read length
+                    return null;
+                } else {
+                    if (useAsynchronousIO) {
+                        byte[] buffer = new byte[readLength];
+                        mCompressedInputStream.read(buffer);
```

Comment:
There's already a method in BlockCompressedInputStream that does this called readBytes but it's a private static method.  (it's also duplicated 4 times in different places in our code base...) It might make sense to expose it as a public method in IOUtil and and then use that here instead of read.